kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/data/RecordConverter.java

```diff
@@ -0,0 +1,508 @@
+    }
+    throw new UnsupportedOperationException("Unsupported type: " + type.typeId());
+  }
+
+  protected GenericRecord convertStructValue(
+      Object value,
+      StructType schema,
+      int parentFieldId,
+      SchemaUpdate.Consumer schemaUpdateConsumer) {
+    if (value instanceof Map) {
+      return convertToStruct((Map<?, ?>) value, schema, parentFieldId, schemaUpdateConsumer);
+    } else if (value instanceof Struct) {
+      return convertToStruct((Struct) value, schema, parentFieldId, schemaUpdateConsumer);
+    }
+    throw new IllegalArgumentException("Cannot convert to struct: " + value.getClass().getName());
+  }
+
+  private GenericRecord convertToStruct(
+      Map<?, ?> map,
+      StructType schema,
+      int structFieldId,
+      SchemaUpdate.Consumer schemaUpdateConsumer) {
+    GenericRecord result = GenericRecord.create(schema);
+    map.forEach(
+        (recordFieldNameObj, recordFieldValue) -> {
+          String recordFieldName = recordFieldNameObj.toString();
+          NestedField tableField = lookupStructField(recordFieldName, schema, structFieldId);
+          if (tableField == null) {
+            // add the column if schema evolution is on, otherwise skip the value,
+            // skip the add column if we can't infer the type
+            if (schemaUpdateConsumer != null) {
+              Type type = SchemaUtils.inferIcebergType(recordFieldValue, config);
+              if (type != null) {
+                String parentFieldName =
+                    structFieldId < 0 ? null : tableSchema.findColumnName(structFieldId);
+                schemaUpdateConsumer.addColumn(parentFieldName, recordFieldName, type);
+              }
+            }
+          } else {
+            result.setField(
+                tableField.name(),
+                convertValue(
+                    recordFieldValue,
+                    tableField.type(),
+                    tableField.fieldId(),
+                    schemaUpdateConsumer));
+          }
+        });
+    return result;
+  }
+
+  private GenericRecord convertToStruct(
+      Struct struct,
+      StructType schema,
+      int structFieldId,
+      SchemaUpdate.Consumer schemaUpdateConsumer) {
+    GenericRecord result = GenericRecord.create(schema);
+    struct
+        .schema()
+        .fields()
+        .forEach(
+            recordField -> {
+              NestedField tableField = lookupStructField(recordField.name(), schema, structFieldId);
+              if (tableField == null) {
+                // add the column if schema evolution is on, otherwise skip the value
+                if (schemaUpdateConsumer != null) {
+                  String parentFieldName =
+                      structFieldId < 0 ? null : tableSchema.findColumnName(structFieldId);
+                  Type type = SchemaUtils.toIcebergType(recordField.schema(), config);
+                  schemaUpdateConsumer.addColumn(parentFieldName, recordField.name(), type);
+                }
```

Comment:
CMIIW but I'm wondering if there is a bug here? When we enter this branch, although we update the table schema to add a new column, we don't seem to add the value of that new column to the `(GenericRecord) result`. 

I feel like this test _should_ pass but currently it does not: 
```java
  @Test
  public void testNewColumn() {
    final org.apache.iceberg.Schema table_schema = new org.apache.iceberg.Schema(
            NestedField.required(20, "i", IntegerType.get()));
    final Table table = mock(Table.class);
    when(table.schema()).thenReturn(table_schema);

    final Schema connect_schema =
            SchemaBuilder.struct()
                    .field("i", Schema.INT32_SCHEMA)
                    .field("l", Schema.INT32_SCHEMA);
    final Struct data = new Struct(connect_schema).put("i", 1).put("l", 2);

    final RecordConverter converter = new RecordConverter(table, config);
    final SchemaUpdate.Consumer consumer = new SchemaUpdate.Consumer();
    final Record record = converter.convert(data, consumer);

    // should add column "l"
    assertThat(consumer.empty()).isFalse();

    GenericRecord rec = (GenericRecord) record;
    assertThat(rec.getField("i")).isEqualTo(1);
    assertThat(rec.getField("l")).isEqualTo(2); // this assertion fails, expected=2, actual=null
  }
```

Reply:
The value isn't added to the record until it exists in Iceberg. This will return the record to `IcebergWriter` without the value but with the requested schema change. `IcebergWriter` will apply the schema change and reconvert the record, at which point the value should be present in the record.